### DISCUSSION TITLE
State per component on react native apps using react navigation

### DISCUSSION THREAD

**Discussion by midnightblurz** [pmndrs/zustand#1923] [Q&A] [answered]
Hello! 

So i am having a tough time finding the best solution to this problem...

Lets say we are using react native navigation with zustand. We have a page to view the users profile, from that user profile we can navigate to another user profile (navigating to same route but with different parameters), if we created a global store, when we navigate to the users profile we request & load the users data, when we navigate from that user profile to the next user profile -> we do the same again, but if we go back to the previous user, we now have the state of the new route ( we lose the navigation state persistance). As i found based on the docs, we can create a state per component.

`const [userStore] = useState(() => create(useUserStore));
`

This solves the problem, but what if we have to navigate to some other routes which are based on the current user, how do we share the state? I tried passing the state to navigation params of the related route like this:

```
navigation.navigate("UserDetails", {
  userStore: userStore,
});

```

Then in the UserDetails route:

```
  const { userDetails } = route.params.watchStore(state => ({
    userDetails: state.userDetails,
  }), shallow);
```

But i don't think this is the most clean and best approach to share the state, any idea how to make this better?

**Comment by dbritto-dev** [comment]
@midnightblurz you can try with this approach https://github.com/pmndrs/zustand/discussions/1902#discussioncomment-6305038

**Comment by midnightblurz** [comment]
> @midnightblurz you can try with this approach [#1902 (reply in thread)](https://github.com/pmndrs/zustand/discussions/1902#discussioncomment-6305038)

I found another solution using context providers and nested stack navigators


So basically we create an instance to the store in our context provider:

```
const UserState = createContext({
  userStore: null,
});

export const useMyListState = () => {
  return useContext(UserState);
};


export default function UserContext({ children }) {
  const [userStore] = useState(() => create(useUserStore));

  return (<UserState.Provider value={{
    userStore: userStore,
  }}>
    {children}
  </UserState.Provider>);
}

```

Then we just create a new stack navigator and wrap it using our context

```
<UserContext>
  <Stack.Navigator initialRouteName="UserStack">
      <Stack.Screen name="UserPage"  component={UserPage} />
      <Stack.Screen name={"UserDetails"} component={UserDetails}/>
  </Stack.Navigator>
</UserContext>
```

In our main Stack.Navigator we use the stack above just like a regular route

```
<Stack.Screen name={"UserStack"} component={UserStack} options={{
  title: null,
  headerShown: false,
  orientation: "portrait",
}} />

```

Then all the pages inside the UserStack will have there own instance of userStore and can access the store simply calling:

`const { userStore } = useUserState();
`

This approach is good enough for me, simple and clean.

**Reply by dbritto-dev** [reply, marked as answer]
@midnightblurz yeah, that works too
